File: RemoveNthNodeFromLL.py
"""
Time Complexity : O(n)
Space Complexity : O(1)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        if not head:
            return head
        # Taking dummy varibale to be able to handle the case when we need
        # to delete the first node
        dummy = ListNode(-1)
        dummy.next = head
        slow = dummy
        fast = dummy
        count = 0
        # Increment fast pointer till it reached count equal n
        while count <= n:
            fast = fast.next
            count += 1
        # After fast pointer reaches end increment slow pointer
        while fast != None:
            fast = fast.next
            slow = slow.next
        # Now we've reached the correct position so unlink and delete
        slow.next = slow.next.next
        return dummy.next

File: remove-nth-node-from-LL.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
// Time Complexity :O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode :Yes
// Any problem you faced while coding this :No
//used two pointer approach Slow and fast pointer to create the space of n between two pointer 
//and then increated both fast and slow at same rate till fast !=null
//linked the slow.next =slow.next.next;
//and then return head using dummy.next
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
      if(head==null){
          return null;
      }  
      ListNode dummy=new ListNode (-1);
      dummy.next=head;
      ListNode slow=dummy;
      ListNode fast=dummy;
      int count=0;
      while(count<=n){
          fast=fast.next;
          count++;
      }
      while(fast !=null){
          slow=slow.next;
          fast=fast.next;
      }
      slow.next=slow.next.next;
      return dummy.next;
    }
}
